include/CL/cl_ext_pocl.h

```diff
@@ -60,42 +60,89 @@ typedef CL_API_ENTRY cl_int
+   a device's memory, but might not be necessarily so, as long as the address
+   range of the buffer remains constant.
+
+   The address is guaranteed to remain the same until the buffer is freed and
+   the address can be queried via clGetMemObjectInfo().
+
+   The device-specific buffer content updates must be still performed by
+   implicit or explicit buffer migrations performed by the runtime or the
+   client code. If any of the devices in the context does not support
+   such allocations, an error (CL_INVALID_VALUE) is returned.
+
+*/
+#define CL_MEM_DEVICE_ADDRESS_EXT (1ul << 31)
+
+/* Experimental multi-device allocation support flags. */
+
+/* If combined with CL_MEM_DEVICE_ADDRESS_EXT, each device can have their
+   own (fixed) device-side address and copy of the created buffer which are
+   synchronized implicitly. The main difference to a default cl_mem allocation
+   is then that the addresses are queriable with CL_MEM_DEVICE_PTRS_EXT and
+   the per-device address is guaranteed to be the same for the entire-lifetime
+   of the cl_mem.
 
-/* TODO: We need also platform/runtime extension due to the new buffer
-   creation flags. */
+   Otherwise, each compatible device in the context can access a shared
+   allocation in one of the device's global memory using the same address. */
+#define CL_MEM_DEVICE_PRIVATE_EXT (1ul << 30)
 
-/* clCreateBuffer(): A new cl_mem_flag CL_MEM_PINNED:
+/* clGetMemObjectInfo(): A new cl_mem_info type CL_MEM_DEVICE_PTR_EXT:
 
-   This flag specifies that the buffer must be persistently allocated
-   in the device's physical memory for its lifetime. That is, the buffer's
-   device address will remain the same and the space is reserved until
-   the buffer is freed. The device-specific buffer content updates are
-   still performed by implicit or explicit buffer migrations performed by
-   the runtime or the client code. If any of the devices in the context
-   does not support pinning, an error (TO DEFINE) is returned.
+   Returns the device address (as a void*) for a buffer allocated with
+   CL_MEM_DEVICE_ADDRESS_EXT. If the buffer was not created with the flag,
+   returns CL_INVALID_MEM_OBJECT.
 */
-#define CL_MEM_PINNED (1ul << 31)
+#define CL_MEM_DEVICE_PTR_EXT 0xff01
 
-/* clGetMemObjectInfo(): A new query CL_MEM_DEVICE_PTR:
+typedef void *cl_mem_device_address_EXT;
 
-Returns a list of pinned device addresses for a buffer allocated
-with CL_MEM_PINNED. If the buffer was not created with CL_MEM_PINNED,
-returns CL_INVALID_MEM_OBJECT.
+/* Returns the device-address pairs for a buffer allocated with
+   CL_MEM_DEVICE_ADDRESS_EXT | CL_MEM_DEVICE_PRIVATE_EXT.
 */
-#define CL_MEM_DEVICE_PTRS 0xff01
+#define CL_MEM_DEVICE_PTRS_EXT 0xff02
 
-typedef struct _cl_mem_pinning
-{
-  cl_device_id device;
-  void *address;
-} cl_mem_pinning;
+  typedef struct _cl_mem_device_address_pair_EXT
+  {
+    cl_device_id device;
+    cl_mem_device_address_EXT *address;
```

Comment:
this needs to be `cl_mem_device_address_EXT address;` otherwise this ends up as a pointer to a pointer or `void **address` but I think once this is a int type this shouldn't matter anymore. Just something I hit while implementing it again.